ARCs/arc-0024.md

```diff
@@ -0,0 +1,485 @@
+place need some way of discovering them. They also need to know that
+the versions of both Application and Application Spec match.
+
+## Goals
+
+This section lists the goals and needs the Application Specification intends 
+to achieve and fulfill.
+
+1. Support tools and IDEs to automatically generate Application proxy classes, including ABI methods, base 'raw' methods and state variables.
+2. The Application Specification format must be human readable and machine readable.
+3. The Application Specification format must be extensible, to allow for inclusion of new ideas,
+such as discovery tags, NFT specifications, and so on.
+4. Support discovery by establishing a link between Applications and their corresponding
+Application Specification files.
+5. Composability: Applications may inherit behaviour from each other in different ways, such as
+implementing interfaces, composing and wrapping other application, direct single inheritance and so on.
+
+The below sections address the above goals, each adding some discussion
+and relevant specification details.
+
+### **1. IDE Support**
+
+There are two general ways of interacting with an Application. One is by making an
+application call transaction and the other is by reading global or local state.
+The state variables are named.
+
+An imaginary Application called SugarSupplierContract offers
+a price for refined sugar to the public. This can be read from the
+global state variable "SugarPrice." A retailer negotiates a discount
+wholesale price. This is offered to the customer via a local state variable
+called "YourPrice." The Application also offers a method to calculate 
+the estimated shipping price for a given volume of sugar.
+
+The retailer would like to be integrate with this SugarSupplierContract
+using the development tool of their choice.
+
+The retailer, in this scenario already has the Application Specification
+available, so they give this to their IDE and the following proxy class is
+generated. The below is just a C# example.
+
+```cs
+   public abstract class SugarSupplierContract : SmartContractReference
+    {
+        [Storage(StorageType.Global)]
+        public ulong SugarPrice;
+
+        [Storage(StorageType.Local)]
+        public ulong YourPrice;
+
+  
+        public abstract ulong ShippingCost(ulong VolumeInKilos);
+    } 
+```
+
+Proxies can support server or dapp (frontend) calls to an Algorand node
+or even contract to contract invocations for compilers.
+
+Even without the IDE to tooling support, a dapp author
+still needs to understand how the SugarSupplierContract is intended
+to be used, so information must be provided somehow.
+
+To achieve this goal, this ARC document proposes that the
+Application Specification include a single ARC-4 Contract
+description, and that the ARC-4 Contract section be updated to
+include locals and globals:
+
+```typescript
+interface Contract {
+  /** A user-friendly name for the contract */
```

Comment:
This section seems to show its heritage from when it was being proposed as a change to ARC4.  I'd rather see this presented without all of the details around `Contract`, and instead something like, "This ARC proposes two optional additional keys to the Contract" schema of ARC4: `globals` and `locals`.  Or, perhaps a cleaner why to plan for the future is to only propose one new key, `ARC24`, which would itself have subkeys `global` and `local`.

I think it's reasonable (though perhaps not explicitly stated in ARC4) to expect that adding new keys to an existing format is fine. So this would be a compatible addition to the Contract Schema, and ARC24 compliant explorers/etc could use it.